File: js/monad.js
/*
  All @Credits to Martin Novak @ https://github.com/MeetMartin.
  This whole project started with the very recommandable short of 7urtle Javascript:
  https://www.youtube.com/watch?v=LITiJmyiOt0
  JavaScript functional programming.
  Monad under 60 seconds
*/

/*
  This is a minimal implementation of a functor
  It simply has a "of"-function that accepts a value
  and returns an object that has "map".
*/
const Functor = {
  of: value => ({
    value: value,
    inspect: () => console.log(`Value: ${value}`),
    map: fn => Functor.of(fn(value))
  })
};

// To turn our functor into a monad we just add the flatMap function.
const Monad = {
  of: value => ({
    __proto__: Functor.of(value),
    flatMap: fn => fn(value)
  })
};


/*
  Every monad is a functor. A functor is just a wrapper about some value:
*/
Functor.of('string');
Functor.of(['array']);
Functor.of(5);

Functor
  .of(5)
  .map(a => a + 1);
// => Functor(6)

/*
  To be a functor, a functor has to implement "map".
  "map" applies a function to the value inside of the functor.
*/

/*
  If a functor maps a function that returns a monad
  you would have a monad inside of a monad.
*/
Monad
  .of('Turtle')
  .map(a => Monad.of(a + 's'));
// => Monad(Monad('Turtles'))


/*
  Instead a monad needs to implement "flatMap".
  "flatMap" flattens the result so instead of a monad inside of a monad you just have one monad.
*/
Monad
  .of('Turtle')
  .flatMap(a => Monad.of(a + 's'));
// => Monad('Turtles')

// Let's apply our monad...
Monad.of(5).flatMap(a => Monad.of(a + 1)).inspect();

File: java/demo/src/main/java/com/example/Monad.java
package com.example;

import java.util.function.Function;

import com.example.spec.Inspectable;
import com.example.spec.IsMonad;

/**
 * To turn our functor into a monad we just add the flatMap function.
 */
public class Monad<T> implements IsMonad<T>, Inspectable {

    private T value;

    /**
     * Compute a monad with the given value
     *
     * @param <T> Type of the monad
     * @param value monad value
     * @return Monad with the given value
     */
    public static <T> Monad<T> of(T value) {

        final Monad<T> monad = new Monad<>();
        monad.value = value;

        return monad;

    }

    /**
     * Prints the monad value to the console.
     */
    @Override
    public void inspect() {

        System.out.println("Monad(" + value + ")");

    }

    /**
     * Applies a function to the value of the monad.
     *
     * @param fn Function which returns the U type as the given type T
     * @return Return a monad with the applied value
     */
    @Override
    public <U> Monad<U> map(Function<T, U> fn) {

        return of(flatMap(fn));

    }

    /**
     * Applies a function to the value of the monad.
     *
     * @param fn Function which returns the U type given the type T
     * @return Return the applied value
     */
    @Override
    public <U> U flatMap(Function<T, U> fn) {

        return fn.apply(value);

    }

    /**
     * Get the monad value.
     *
     * @return The monad value.
     */
    @Override
    public T getValue() {

        return flatMap(Function.identity());

    }
}
